src/docs/source_links.rs

```diff
@@ -3,156 +3,87 @@ use crate::{
+
+        let url_pattern = match &project_config.repository {
+            Repository::GitHub { user, repo } => Some((
+                format!(
+                    "https://github.com/{}/{}/blob/v{}/{}#L",
+                    user, repo, project_config.version, path_in_repo
+                ),
+                "-L".to_string(),
+            )),
+            Repository::GitLab { user, repo } => Some((
+                format!(
+                    "https://gitlab.com/{}/{}/-/blob/v{}/{}#L",
+                    user, repo, project_config.version, path_in_repo
+                ),
+                "-".to_string(),
+            )),
+            Repository::BitBucket { user, repo } => Some((
+                format!(
+                    "https://bitbucket.com/{}/{}/src/v{}/{}#lines-",
+                    user, repo, project_config.version, path_in_repo
+                ),
+                ":".to_string(),
+            )),
+            Repository::Custom { .. } | Repository::None => None,
+        };
+
+        SourceLinker {
+            line_starts: if url_pattern.is_some() {
+                line_starts(&module.src)
+            } else {
+                vec![]
+            },
+            url_pattern,
+        }
     }
-}
-
-// Creates links to the BitBucket repository with the #lines-M:N anchor syntax for linking directly
-// to a range of line numbers
-struct BitBucketSourceLinks {
-    codespan_file: codespan_reporting::files::SimpleFile<String, String>,
-    user: String,
-    repo: String,
-    version: String,
-    relative_path: String,
-}
-
-impl SourceLinks for BitBucketSourceLinks {
-    fn url(&self, location: &SrcSpan) -> Option<String> {
-        let (start_line, end_line) = get_lines(&self.codespan_file, location);
-        Some(format!(
-            "https://bitbucket.com/{}/{}/src/v{}/{}#lines-{}:{}",
-            &self.user, &self.repo, &self.version, &self.relative_path, start_line, end_line,
-        ))
+    pub fn url(&self, span: &SrcSpan) -> String {
+        match &self.url_pattern {
+            Some((base, line_sep)) => {
+                let start_line = line_no(&self.line_starts, span.start);
+                let end_line = line_no(&self.line_starts, span.end);
+                format!("{}{}{}{}", base, start_line, line_sep, end_line)
+            }
+
+            None => "".to_string(),
+        }
     }
 }
 
-fn get_lines(
-    codespan_file: &codespan_reporting::files::SimpleFile<String, String>,
-    location: &SrcSpan,
-) -> (usize, usize) {
-    let start_line = codespan_file
-        .line_index((), location.start)
-        .unwrap_or_default()
-        + 1;
-    let end_line = codespan_file
-        .line_index((), location.end)
-        .unwrap_or_default()
-        + 1;
-
-    (start_line, end_line)
+// get all the line start byte offsets for a str
+fn line_starts(src: &str) -> Vec<usize> {
+    std::iter::once(0)
+        .chain(src.match_indices('\n').map(|(i, _)| i + 1))
```

Comment:
Does this work when non `\n` line endings are used?

Reply:
Oh good question, if the line endings are `\n`(posix) or `\r\n`(win) it should be fine. Will not if they are just `\r`(vintage).

It's copy-pasta from codespan: https://github.com/brendanzab/codespan/blob/master/codespan-reporting/src/files.rs#L251

Reply:
That's probably ok I think? What do you reckon?

Reply:
I think it's probably fine, kinda deferring judgement to codespan.

Reply:
Let's go then :)